### PR TITLE
Jack

### PR DESCRIPTION
**Pull request by jackraymond**
Major refactoring.

Mostly documentation, and adding some options (verbose, max_num_emb, use_cache, etc) to main options.
Removed some verbose statements.
Error with client deletion during caching data corrected.
Simplified/standardized embed_loops.py and embed_square_lattice.py ; including __main__
Note that an additional dependency on minorminer is introduced, for now this requires specifically the jack branch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,38 +1,73 @@
...
+```
+
+For generation of orbits related to Figures 13-16
+```
+python -m example3_1_tafm_get_orbits
+```
+
 ## Ferromagnetic loop example
+
+Requires 300 sequential QPU-job submissions.
```
Comment (jackraymond) [inline README.md:48]: Should add: 'For generation of Figure XX like the other cases'

File: tutorial_code/embed_loops.py [hunk 2]
```diff
@@ -57,49 +61,48 @@ def embed_loops(sampler, L, try_to_load=True, raster_breadth=2):
...
-            print("Finding embedding via raster embedding search instead.")
-
     # Check if the target graph has enough nodes
-    if G.number_of_nodes() > A.number_of_nodes():
-        raise EmbeddingError(
-            f"Source graph has {G.number_of_nodes()} nodes, "
-            f"which exceeds the target graph's {A.number_of_nodes()} nodes."
-        )
+    if not subgraph_embedding_feasibility_filter(S=G, T=A):
+        raise ValueError(f'Embedding {S} on {T} is infeasible')
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/embed_loops.py]: I think it should be G and A in the warning message as we don't have access to S and T.

File: tutorial_code/embed_loops.py [hunk 3]
```diff
@@ -13,42 +13,46 @@
...
+                                               embeddings_to_ndarray,
+                                               raster_breadth_subgraph_lower_bound,
+                                               raster_breadth_subgraph_upper_bound,
+                                               subgraph_embedding_feasibility_filter)
 
-def embed_loops(sampler, L, try_to_load=True, raster_breadth=2):
-    """
-        Attempts to expand an independent set by replacing subsets of size 'greed' with larger independent sets.
-        The function iteratively tries to improve the given independent set by exploring the neighborhood of subsets.
+def embed_loops(sampler, L, use_cache=True, raster_breadth=None, **re_kwargs):
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/embed_loops.py]: Recommand type hinding
```
def embed_loops(sampler: DWaveSampler, L: int, use_cache: bool = True,
                raster_breadth: int = None, **re_kwargs) -> np.ndarray:
```

File: tutorial_code/embed_square_lattice.py [hunk 4]
```diff
@@ -47,46 +40,78 @@ def embed_square_lattice(sampler, L, try_to_load=True, **kwargs):
...
+        embeddings = np.loadtxt(cache_filename, dtype=int)
+        if embeddings.ndim==1:
+            embeddings = embeddings[np.newaxis,:]
+        print(f'Loaded {cache_filename}')
+        return embeddings, bqm
+    else:
+        G = dimod.to_networkx_graph(bqm)
+        A = sampler.to_networkx_graph()
+        if not subgraph_embedding_feasibility_filter(S=G, T=A):
+            raise ValueError(f'Embedding {S} on {T} is infeasible')
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/embed_square_lattice.py]: Similar variable naming issues as the previous file.

File: tutorial_code/embed_square_lattice.py [hunk 5]
```diff
@@ -47,46 +40,78 @@ def embed_square_lattice(sampler, L, try_to_load=True, **kwargs):
...
+                             '\nModifying the source (lattice) and target '
+                             '(processor), or raster_embedding_search arguments '
+                             'such as timeout may resolve the issue.')
+        
+
+    if use_cache:
+        try:
+            os.makedirs('cached_embeddings/', exist_ok=True)
+            np.savetxt(cache_filename, embeddings, fmt='%d')
+        except:
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/embed_square_lattice.py]: Maybe consider a more foused exceptions in this case:
```
except OSError as e:
    warnings.warn(f'Embedding cache files could not be created: {e}')
    ```

File: tutorial_code/example0_1_orbits.py [hunk 6]
```diff
@@ -43,49 +43,55 @@ def make_bqm():
...
     """
-    # Make the four-qubit BQM and compute its orbits.
+
     bqm = make_bqm()
     signed_bqm = orbits.make_signed_bqm(bqm)
     signed_qubit_orbits, signed_coupler_orbits = orbits.get_bqm_orbits(signed_bqm)
     qubit_orbits, coupler_orbits, qubit_orbits_opposite, coupler_orbits_opposite = \
         orbits.get_orbits(bqm)
+    if verbose:
+        print('Signed qubit orbits: {signed_qubit_orbits}')
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/example0_1_orbits.py]: Should use f-tring:
 `print(f'Signed qubit orbits: {signed_qubit_orbits}')`

File: tutorial_code/example0_1_orbits.py [hunk 7]
```diff
@@ -43,49 +43,55 @@ def make_bqm():
...
+                print(f'Coupler orbit {i} behavior opposite to coupler orbit {coupler_orbits_opposite[i]}.')
 
     plt.rc('font', size=12)
-    plt.figure(figsize=(11, 5), dpi=80)  
+    fig = plt.figure(figsize=(11, 5), dpi=80)
+    fig.canvas.manager.set_window_title('Figure 4: Orbits of signed and original Ising model')
+    
     options = {'node_size': 600, 'width': 4}  
 
     Gnx = orbits.to_networkx_graph(qubit_orbits, coupler_orbits)
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/example0_1_orbits.py:105]: Maybe consider excepton handling:
```
try:
    Gnx = orbits.to_networkx_graph(qubit_orbits, coupler_orbits)
except AttributeError as e:
    raise RuntimeError(f"Error in generating networkx graph: {e}")
```

File: tutorial_code/example0_1_orbits.py [hunk 8]
```diff
@@ -43,49 +43,55 @@ def make_bqm():
...
+                print(f'Qubit orbit {i} has behavior opposite to qubit orbit {qubit_orbits_opposite[i]}.')
 
-    for i in range(len(coupler_orbit_sizes)):
-        if coupler_orbit_sizes[i] > 0 and coupler_orbit_sizes[coupler_orbits_opposite[i]] > 0:
-            print(f'Coupler orbit {i} behavior opposite to coupler orbit {coupler_orbits_opposite[i]}.')
+        for i in range(len(coupler_orbit_sizes)):
+            if coupler_orbit_sizes[i] > 0 and coupler_orbit_sizes[coupler_orbits_opposite[i]] > 0:
+                print(f'Coupler orbit {i} behavior opposite to coupler orbit {coupler_orbits_opposite[i]}.')
 
     plt.rc('font', size=12)
```
Comment (AndyZzzZzzZzz) [inline tutorial_code/example0_1_orbits.py:101]: We could onsider modularize this part as a separate visualization function for code clarity.

File: tutorial_code/embed_square_lattice.py [hunk 9]
```diff
@@ -14,25 +14,18 @@
 #
 import os
 import dimod
+import warnings
 import numpy as np
+import time  # temporary
```
Comment (jackraymond) [inline tutorial_code/embed_square_lattice.py]: Remove, I think this was for debugging.

File: tutorial_code/example1_2_fm_loop_correlations.py [hunk 10]
```diff
@@ -179,73 +179,93 @@ def run_experiment(param, shim, stats, embeddings, _alpha_Phi=0., _alpha_J=0.):
...
-    paper_plots_example1_2(all_couplings=stats['all_couplings'], 
+    paper_plots_example1_2(all_couplings=stats['all_couplings'],
                            all_fbos=stats['all_fbos'],
                            mags=stats['mags'],
                            frust=stats['frust'])
 
 
-def main(sampler_type='mock', model_type=None, num_iters=300, num_iters_unshimmed_flux=100, num_iters_unshimmed_J=200):
-    """Main function to run example
+def main(solver_name=None, model_type=None, num_iters=300,
```
Comment (jackraymond) [inline tutorial_code/example1_2_fm_loop_correlations.py]: Change model_type to coupling=-0.2

**Pull request merged**